src/Strategy/Gradle.hs

```diff
@@ -107,12 +108,29 @@ discover dir = context "Gradle" $ do
 
 -- Run a Gradle command in a specific working directory, while correctly trying
 -- Gradle wrappers.
-runGradle :: (Has Exec sig m, Has Diagnostics sig m) => Path t Dir -> (Text -> Command) -> m BL.ByteString
+runGradle :: (Has ReadFS sig m, Has Exec sig m, Has Diagnostics sig m) => Path Abs Dir -> (Text -> Command) -> m BL.ByteString
 runGradle dir cmd =
-  execThrow dir (cmd (toText dir <> "gradlew"))
-    <||> execThrow dir (cmd (toText dir <> "gradlew.bat"))
+  do
+    walkUpDir dir "gradlew" >>= execThrow dir . cmd . toText
+    <||> (walkUpDir dir "gradlew.bat" >>= execThrow dir . cmd . toText)
     <||> execThrow dir (cmd "gradle")
 
+-- Search upwards in a directory for the existence of the supplied file.
+walkUpDir :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> Text -> m (Path Abs File)
+walkUpDir dir filename = do
+  relFile <- case parseRelFile $ Data.String.Conversion.toString filename of
+    Nothing -> fatal $ toText $ "invalid file name: " <> filename
```

Comment:
[nit]: Since `filename` is `Text`, this `toText` is just `Text -> Text`.
```suggestion
    Nothing -> fatal $ "invalid file name: " <> filename
```